Add instance UUID support

kryksyh opened the pull request:
Implementation of instance UUID support

For more information:
[UUID info](https://forum.audacityteam.org/t/introducing-a-uuid-system-in-audacity-here-s-what-you-need-to-know/128561/10)
[Privacy Policy](https://forum.audacityteam.org/t/changes-to-our-privacy-policy-for-audacity-3-7-2/130309)


QA please verify that:
* If the user declines to create UUID, the UUID is not created and not saved in `audacity.cfg`
* UUID if created, survive settings reset, and is not changed

<!-- Use "x" to fill the checkboxes below like [x] -->

- [x] I signed [CLA](https://www.audacityteam.org/cla/)
- [x] The title of the pull request describes an issue it addresses
- [x] If changes are extensive, then there is a sequence of easily reviewable commits
- [x] Each commit's message describes its purpose and effects
- [x] There are no behavior changes unnecessary for the stated purpose of the PR

Recommended:
- [x] Each commit compiles and runs on my machine without known undesirable changes of behavior

DavidBailes commented:
@LWinterberg . In https://forum.audacityteam.org/t/introducing-a-uuid-system-in-audacity-here-s-what-you-need-to-know/128561
it states that "It’s important to note that UUIDs, while random, are considered ‘personal information’ under GDPR."
However, in this PR, both on the opt in/out dialog, and in application preferences it states that Audacity does not store any personal information.
I'm confused.

DavidBailes commented:
Anonimous is not the correct spelling of anonymous.

saintmatthieu commented on libraries/lib-cloud-audiocom/NetworkUtils.cpp:
```diff
@@ -25,6 +26,12 @@ namespace audacity::cloud::audiocom
 
+namespace optional_headers {
+
+const std::string InstanceId = "x-audacity-instance-id";
```
Just curious: what's the leading `x` for ?

kryksyh replied:
it was asked from the backend team, not sure about that

kryksyh commented on src/update/UpdateManager.cpp:
```diff
@@ -70,7 +79,9 @@ void UpdateManager::Start(bool suppressModal)
 
-            notice.ShowModal();
+            int result = notice.ShowModal();
+            SendAnonimousUsageInfo->Write(result != wxNO);
```
Yes 😅

kryksyh commented on src/update/UpdateManager.h:
```diff
@@ -53,11 +53,13 @@ class UpdateManager final : public wxEvtHandler
     /// Scheduling update time for avoiding multiplying update notifications.
     bool IsTimeForUpdatesChecking();
 
+    std::string GetUpdatesUrl();
```
Yes, it should be const

kryksyh commented on libraries/lib-cloud-audiocom/OAuthService.cpp:
```diff
@@ -403,6 +411,8 @@ void OAuthService::Authorize(std::string_view email,
    request.setHeader(common_headers::ContentType, common_content_types::ApplicationJson);
    request.setHeader(common_headers::Accept, common_content_types::ApplicationJson);
 
+   SetOptionalHeaders(request);
```
We are setting these headers only for requests sent to the audio.com backend. Avatars and project data requests are sent to the storage, and they do not require those.

saintmatthieu replied:
Ok, so I should not expect any `SetOptionalHeaders` outside the lib-cloud-audiocom module.
Still I find it hard to figure out which requests should have the UUID and which not. Is there an explicit list someplace, or anything to help code readers?

kryksyh replied:
> so I should not expect any SetOptionalHeaders outside the lib-cloud-audiocom module.
yes
> Is there an explicit list someplace, or anything to help code readers?
No, I don't think so

kryksyh commented on src/prefs/ApplicationPrefs.cpp:
```diff
@@ -108,6 +131,7 @@ bool ApplicationPrefs::Commit()
    ShuttleGui S(this, eIsSavingToPrefs);
    PopulateOrExchange(S);
    DefaultUpdatesCheckingFlag->Invalidate();
+   SendAnonimousUsageInfo->Invalidate();
```
After the preferences window is closed, the configuration file is updated, but the Option objects may contain cached data that needs to be updated.

kryksyh commented on src/update/UpdateManager.cpp:
```diff
@@ -30,14 +32,20 @@
+   void UpdatePrefs() override;
+};
+
+static std::optional<UpdaterPrefsListener> prefsListener;
```
That's a good point, it makes sense.

kryksyh commented:
@DavidBailes thanks for the feedback!

> However, in this PR, both on the opt in/out dialog, and in application preferences it states that Audacity does not store any personal information.

I have updated the message to be more correct: `Audacity does not collect any personally identifiable information about you.`

> Anonimous is not the correct spelling of anonymous.

Indeed!

> In Application preferences, the Send anonimous (anonymous) usage info check box is inside an "update notifications" group box, which having nothing to do with updates.

Group name was renamed to "Network communications"

> Also, this check box does not have an access key.

I fixed this too.

DavidBailes commented:
This is comment is about wording, rather than code, and so is probably principally directed at @LWinterberg  This is part of the dialog about updates and sending  anonymous data:
![image](https://github.com/user-attachments/assets/c71554a7-98b8-438f-8e92-ff775d3d6122)

In plain language, the question being asked of the user is whether they are happy to send anonymous data. That Audacity is using a UUID is a technical detail that explains for those who are interested how the data is kept anonymous. But Audacity could generate an UUID and send no data.
A button with the name "Disable sending anonymous data" rather than "Disable UUID" would be much more easily understood by most users.

LWinterberg commented:
> A button with the name "Disable sending anonymous usage data" rather than "Disable UUID" would be much more easily understood by most users.

It unfortunately also would be incorrect as update checking would remain enabled.

DavidBailes commented:
> > A button with the name "Disable sending anonymous usage data" rather than "Disable UUID" would be much more easily understood by most users.
> 
> It unfortunately also would be incorrect as update checking would remain enabled.

You mean because update checking also provides anonymous usage data?

dozzzzer commented:
Tested

Tantacrul commented:
@DavidBailes We could change it to 'Don't send anonymous UUID', however, I kind of feel it won't make sense to anyone until they've read the description at the top, so we went for this more simplified explanation.

Although, in both cases, the option won't make sense to just about anyone if they don't read the text.

Also, just in case you were wondering - we decided to keep these notifications together because we think it'd get super annoying to have multiple back-to-back dialogs. Most likely, the user would just spam the 'OK' button if we did that and be even less inclined to read each dialog.

So, I guess I'm pretty much saying, I think the button should remain as is. There's also the issue of space / the buttons being really massive too.

Thoughts?

DavidBailes commented:
> I guess I'm leaning towards leaving the button as is. There's also the issue of space / the buttons being really massive too.
> 
> Thoughts?

@Tantacrul . Having the two options in one dialog makes sense.
Often opting in/out of options such as these is done using check boxes, rather than buttons. This removes the issue of problems of size of buttons whilst trying to make their purpose clearly understandable.
My understanding is that the new option you are introducing is whether the user is happy with sending the information of how often this particular copy of Audacity is being used, where this information is being sent anonymously.
What I definitely don't understand is the section in the privacy document relating to Audio.com:

> Audio.com integration insights – for users already registered with Audio.com, this system will help us understand how often Audacity users take advantage of Audio.com. We will not be sharing any other kinds of information with Audio.com, apart from the UUID itself.

If a user is signed into Audio.com, then they already know how often that user is using the service don't they. So I don't know what the purpose is. Linking an UUID with personal information so that the UUID is no longer anonymous doesn't sound like a good idea to me.

Currently, if a user upgrades to 3.7.2 from a previous version, then this dialog doesn't open. I'm not sure whether this is by design or not. But if it did, then I think the user would be puzzled about why they are being asked about update checking again.

DavidBailes commented:
> To your question, if you're on Audio.com, you're obviously already signed in with your email and password. The UUID doesn't provide any additional personally identifying info because it's a randomly generated number. All it does is give us a better sense of overall retention and helps us understand the overall % of Audacity sessions that include usage of Audio.com.

@Tantacrul  It would be helpful if you could explain exactly how you are going to calculate the overall % of Audacity sessions that include usage of Audio.com.

Tantacrul commented:
Let me rephrase - the overall % of those who opt in to the UUID. We can't know about those that don't.

DavidBailes commented:
> Let me rephrase - the overall % of those who opt in to the UUID. We can't know about those that don't.

So is this right: you are wanting to calculate, for those who opt in to the UUID, the overall % of Audacity sessions that include the usage of Audio.com?
As I've said above, an explanation of how you are going to calculate this would be helpful.
If, for example, this involved Audio.com being sent the UUID at the start of each session, then that usage information would not be anonymous.

Tantacrul commented:
**One quick point on terminology:**

Regarding the message about UUIDs: Under GDPR, a UUID alone is not classified as 'anonymous' data, even if collected without an IP address (and to clarify, we do not store an IP address with it). Instead, it is considered pseudonymous data. For that reason, we avoid using the term "anonymous" when discussing the UUID. What we do state—accurately and in alignment with GDPR—is that we do not collect any personally identifying information (PII).

Regarding update checking: We do use the term "anonymous" here because, under GDPR, the data collected for update checking is truly anonymous.

**Regarding the message itself:**

The message is accurate for standard Audacity usage. Introducing details about what happens if a user later chooses to register on Audio.com would overcomplicate the message to the point of making it unclear.

Instead, it makes more sense to provide additional details only when a user is considering Audio.com, by including a link to the Audio.com Privacy Policy in all relevant places—which we already do.

Tantacrul commented:
Ah, hold on, I see what you're saying now. The text in this build is including the word 'anonymous' still. That's a mistake.

@kryksyh - we need to revise that. I think the Figma mockup had outdated text in it. I can't seem to get this PR working but I see its merged, so let me grab master and see what the text actually says now and we'll replace the 'anonymous' parts with a correct description.

DavidBailes commented:
In section 6 of the PP, it states that:
We delete: UUID data immediately if you choose to disable it at any point

I understand by that: if I disable the option of sending usage information in Preferences, then all of my prior usage data held by you will be deleted immediately.
Is that actually going to be true?
